redwood-flexbox-engine/src/commonMain/kotlin/app/cash/redwood/FlexboxEngine.kt

```diff
@@ -0,0 +1,2041 @@
+              index = viewIndex,
+            )
+            FlexDirection.Column, FlexDirection.ColumnReverse -> stretchViewHorizontally(
+              node = node,
+              crossSize = flexLine.crossSize,
+              index = viewIndex,
+            )
+            else -> error("Invalid FlexDirection: $flexDirection")
+          }
+          j++
+        }
+        i++
+      }
+    } else {
+      for (flexLine in flexLines) {
+        for (index in flexLine.indicesAlignSelfStretch) {
+          val view = getReorderedChildAt(index)!!
+          when (flexDirection) {
+            FlexDirection.Row, FlexDirection.RowReverse -> stretchViewVertically(
+              node = view,
+              crossSize = flexLine.crossSize,
+              index = index,
+            )
+            FlexDirection.Column, FlexDirection.ColumnReverse -> stretchViewHorizontally(
+              node = view,
+              crossSize = flexLine.crossSize,
+              index = index,
+            )
+            else -> throw IllegalArgumentException("Invalid flex direction: $flexDirection")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Expand the view vertically to the size of the [crossSize] (considering the view margins)
+   *
+   * @param node the View to be stretched
+   * @param crossSize the cross size
+   * @param index the index of the view
+   */
+  private fun stretchViewVertically(node: Node, crossSize: Int, index: Int) {
+    var newHeight = crossSize - node.margin.top - node.margin.bottom
+    newHeight = max(newHeight, node.minHeight)
+    newHeight = min(newHeight, node.maxHeight)
+    val measuredWidth = if (measuredSizeCache != null) {
+      // Retrieve the measured height from the cache because there
+      // are some cases that the view is re-created from the last measure, thus
+      // View#getMeasuredHeight returns 0.
+      // E.g. if the flex container is FlexboxLayoutManager, that case happens
+      // frequently
+      unpackLower(measuredSizeCache!![index])
+    } else {
+      node.measuredWidth
+    }
+    val childWidthSpec = MeasureSpec.from(measuredWidth, MeasureSpecMode.Exactly)
+    val childHeightSpec = MeasureSpec.from(newHeight, MeasureSpecMode.Exactly)
+    node.measure(childWidthSpec, childHeightSpec)
+    updateMeasureCache(index, childWidthSpec, childHeightSpec, node)
+  }
+
+  /**
+   * Expand the view horizontally to the size of the crossSize (considering the view margins)
+   *
+   * @param node      the View to be stretched
+   * @param crossSize the cross size
+   * @param index     the index of the view
+   */
+  private fun stretchViewHorizontally(node: Node, crossSize: Int, index: Int) {
+    var newWidth = crossSize - node.margin.start - node.margin.end
+    newWidth = max(newWidth, node.minWidth)
+    newWidth = min(newWidth, node.maxWidth)
+    val measuredHeight = if (measuredSizeCache != null) {
+      // Retrieve the measured height from the cache because there
```

Comment:
Oooooh this is very interesting. Looks like there’s a case we might want to track where the `View` object gets replaced with another one, and we prefer cache data?